File: client/src/modules/transaction/schemas/transaction.ts
import { PaginationQuery } from "@/lib/pagination";
import { Person } from "@/modules/account/schemas/person";

export enum TransactionType {
  DEPOSIT = 1,
  WITHDRAW = 2,
}

export type Transaction = {
  id: number;
  money: number;
  dateTime: string;
  transactionType: TransactionType;
  account: Person;
  originAccount?: Person;
};

export type TransactionMonthResume = {
  month: string;
  label: string;
  amount: number;
};

export type TransactionQuery = PaginationQuery & {
  transactionDate: Date
  transactionType: TransactionType
}

File: server/knockbankapi/app/schemas/transaction.py
from datetime import datetime as dt, date
from apiflask import Schema
from apiflask.validators import OneOf
from apiflask.fields import Float, Integer, Nested, DateTime, Date, String, List
from marshmallow import validates, ValidationError
from knockbankapi.app.schemas import PaginationQuery, PaginationResponse, PersonBasic


class TransactionQuery(PaginationQuery):
    transactionDate: date = Date()
    transactionType: int = Integer(
        validate=[
            OneOf([1, 2, None], error="Tipo de transação inválida, deve ser 1 ou 2.")
        ],
    )


class TransactionIn(Schema):
    money: float = Float(required=True)

    @validates("money")
    def validate_money(self, money: float, **kwargs):
        if money <= 0:
            raise ValidationError("O valor da transação deve ser maior que zero.")


class TransactionTransfer(TransactionIn):
    accountId: int = Integer(required=True)


class TransactionOut(Schema):
    id: int = Integer()
    money: float = Float()
    dateTime: dt = DateTime()
    transactionType: int = Integer()
    account: dict = Nested(PersonBasic)
    originAccount: dict = Nested(PersonBasic)


class PaginationTransactionOut(PaginationResponse):
    data = List(Nested(TransactionOut))

class TransactionMonthResume(Schema):
    month: str = String()
    label: str = String()
    amount: float = Float()
